### PR TITLE
UI/UX: implements a fully revamped direct message composer

### PR DESCRIPTION
**Pull request by jjaffeux**

### CODE REVIEW

File: assets/stylesheets/common/direct-message-creator.scss [hunk 1]
```diff
@@ -0,0 +1,189 @@
+.direct-message-creator {
```
Comment (pmusaraj) [inline assets/stylesheets/common/direct-message-creator.scss:1]: A general comment (not meant to be a blocker for this PR at all, more of a mental note): we need to extract some of the chat styling into "modular elements", so that it can be more easily reused.

**Review by pmusaraj — APPROVED** [review]
This looks and feels great! Comments in the PR are small nitpicks, nothing major. 

I pulled the PR locally and used it, and one thing that stands out is that the "loading slider" is invoked very often, sometimes it's so fast it feels like an error, like it's too sensitive and quick. 

I'm happy to try this as is on a real server and adjust later. Here's a screencast of what I mean: 

https://user-images.githubusercontent.com/368961/162579945-407f0a15-5688-4f54-8369-51ec9913f351.mp4

File: assets/javascripts/discourse/components/chat-composer.js [hunk 2]
```diff
@@ -119,6 +120,7 @@ export default Component.extend(TextareaTextManipulation, ComposerUploadUppy, {
         title: "chat.upload",
       });
     }
+    this.set("toolbarButtons", toolbarBtns.concat(toolbarExtraButtons));
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply assets/javascripts/discourse/components/chat-composer.js:123]: oops yes indeed :+1:

File: assets/javascripts/discourse/components/chat-user-avatar.js [hunk 3]
```diff
@@ -11,8 +11,11 @@ export default Component.extend({
...
   avatarSize: "tiny",
 
-  @discourseComputed("chat.presenceChannel.users.[]", "user.id")
-  isOnline(users, userId) {
-    return !!users?.findBy("id", userId);
+  @discourseComputed("chat.presenceChannel.users.[]", "user.{id,username}")
+  isOnline(users, user) {
+    return (
+      !!users?.findBy("id", user?.id) ||
+      !!users?.findBy("username", user?.username)
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply assets/javascripts/discourse/components/chat-user-avatar.js:18]: Sometimes chat-user-avayar will receive a component with an id sometimes with a username, both cases exist

File: assets/stylesheets/common/direct-message-creator.scss (hunk 1 above)
Comment (jjaffeux) [reply assets/stylesheets/common/direct-message-creator.scss:1]: Yep, I have already started splitting a lot of components in last weeks, but we certainly need to do much more

File: test/javascripts/acceptance/chat-browse-test.js [hunk 4]
```diff
@@ -191,13 +202,21 @@ acceptance("Discourse Chat - chat browsing no channels", function (needs) {
...
-    await users.expand();
-    await fillIn(".dm-user-chooser input.filter-input", "hawk");
-    await users.selectRowByValue("hawk");
-    await click("button.create-dm");
-    assert.equal(currentURL(), "/chat/channel/75/@hawk");
+
+    assert.equal(
+      currentURL(),
+      `/chat/channel/draft/${encodeURIComponent(
+        I18n.t("chat.direct_message_creator.title")
```
[the remark this answers is not in the document]
Comment (jjaffeux) [reply test/javascripts/acceptance/chat-browse-test.js:209]: Yes mostly this is a kind of edge case to avoid a dedicated route. I'm mixed on this. The big pros is that it avoids a lot of code, but it can feel slightly hacky I agree.

[the remark this answers is not in the document]
Comment (jjaffeux) [reply test/javascripts/acceptance/chat-browse-test.js:209]: I tend to keep it that way for now, but we can think on how to make this better with mark.

**Comment by jjaffeux** [comment]
> I pulled the PR locally and used it, and one thing that stands out is that the "loading slider" is invoked very often, sometimes it's so fast it feels like an error, like it's too sensitive and quick.

Agreed, I made minor changes which should limit this, let's see how it goes

**Comment by jjaffeux** [comment]
Will merge this tomorrow morning, thanks for help!

File: assets/stylesheets/common/direct-message-creator.scss (hunk 1 above)
Comment (pmusaraj) [reply assets/stylesheets/common/direct-message-creator.scss:1]: Awesome, thanks!
